app/forms/schools/add_participant_form.rb

```diff
@@ -6,20 +6,54 @@ class AddParticipantForm
     include ActiveModel::Attributes
     include ActiveModel::Serialization
 
+    STEPS = %i[type details not_implemented].freeze
```

Comment:
I am not sure a `form` is a good place to define steps. I think `wicked` did it in a controller (if I remember correctly), which feels somewhat better. But more importantly...

I much prefer having an explicit action in controller and explicit method on a form for handling a step. It means if steps have a degree of difference between them it is really easy to tell.

Reply:
You're correct - wicked does it in the controller. And I always find it extremely weird and, well, incorrect. Wicked AFAIR is based on a fixed step order which will not work in this case.

The form needs to perform step-related validations, so it would need to reach to the controller to know what step it is on. It should be the form that defines its own flow and order of assignments. Even when we speak about this, we say "mutlistep form", not "multistep controllers" so I'd argue it is form's responsibility to define its steps.

As per explicit actions in the controller - each step is only to assign the attributes to the form. There is absolutely no business logic there - I had those actions as a separate actions before and they were just repeating same code all the time.

Reply:
When actions repeat a lot of code, I prefer to extract the common part into a private method. That makes the differences obvious - say, where do we go after a step fails or succeeds.

If I want to know from this code where we go after a particular step I need to understand what the next step method does (which for now is easy, but it doesn't always need to be). 

And as a general solution, I think it centralises the routing logic in a pretty scary way. Think about our flow for nominating induction tutors - there are several branches that lead into dead ends when particular conditions are met. And the `next_step` method needs to know about those, right?

I remember wicked from a long time ago, and my experience was pretty similar - it did the simple thing we wanted, but we struggled a lot around edge cases. I think if we try to apply this solution to the wider class of problems that we face across this project we will have a very similar experience. And if we don't apply it to any other solution, well, it seems excessively complicated for one flow :( 

For what it's worth... If we really want to go down this road, can we replicate what `apply-for-teacher-training` are doing? They have two wizards (from what I can tell), and it looks like they spent a long time thinking about this problem. 
`ReconfirmDeferredOfferWizard` is probably closer in scope to what we want. They set up a controller action for each step, have a list of steps in the form (well, they call it a wizard, but you know, naming), but the rest looks pretty similar to what you have. The larger one is `OfferWizard` - which has its own `new create edit update` controller per step. A bit much for our need I think.